### PR TITLE
DEV: Add boost to `scipy.special`

### LINKED ISSUE #10317: scipy.stats.nbinom.interval returns wrong result for p=1
When computing confidence intervals with probability of success equals to 1. The function `scipy.stats.nbinom.interval()` doesn't return the expected result.

### Reproducing code example:

```
>>> scipy.stats.nbinom.interval(alpha=0.9, n=10, p=1)
(1e+100, 0.0)
```

### Expected result:

This is what is expected !
```
>>> scipy.stats.nbinom.interval(alpha=0.9, n=10, p=1)
(0.0, 0.0)
```
For instance with Poisson distribution
```
>>> scipy.stats.poisson.interval(alpha=0.9, mu=0)
(0.0, 0.0)
```

### Scipy/Numpy/Python version information:
```
>>> import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info)
1.2.1 1.16.3 sys.version_info(major=3, minor=6, micro=8, releaselevel='final', serial=0)
```

### PR DESCRIPTION
**Pull request by V0lantis**
#### Reference issue
Might Close gh-10317

There are many issues concerning `scipy.special`. The purpose of this PR is to add *boost* library, and progressively **replace** the whole `cdflib` submodule with `boost` functions.

#### What does this implement/fix?

For now, I just tried to implement the **negative binomial cdf** using *boost* implementation, to close gh-10317. I don't know how to replace the link from the actual `nbdtrik` located on my computer's build in `scipy.special._ufuncs.py` to the new *boost* version, located in `scipy.special._boost._nbinom_cdf`. Also, There aren't many tests on `nbdtrik`. the only I found were this [one](https://github.com/scipy/scipy/blob/6925c378bf5766079708420c6caa880612022bca/scipy/special/tests/test_cdflib.py#L375) and this [one](https://github.com/scipy/scipy/blob/6925c378bf5766079708420c6caa880612022bca/scipy/special/tests/test_basic.py#L714)

#### Additional information

I'd like to know : 

- [ ] How to replace the actual implementation `nbdtrik` (IDK if it does actually come from CDFLIB or CEPHES, or what is the difference between those two)  by the boost one
- [ ] Where should I add the new tests in `special`.

@mckib2 Is the right way to do it?

### CODE REVIEW

**Comment by person142** [comment]
> How to replace the actual implementation nbdtrik (IDK if it does actually come from CDFLIB or CEPHES, or what is the difference between those two) by the boost one

You want to look in `functions.json` to answer that question:

https://github.com/scipy/scipy/blob/master/scipy/special/functions.json#L1022

so it comes from cdflib. One thing to be aware of is that you need to make sure the functions from Boost are still exported from `cython_special` (`_generate_pyx.py` does this).

> Where should I add the new tests in special.

As you've noted we don't have any special tests, so I'd add a new file to

https://github.com/scipy/scipy/tree/master/scipy/special/tests

with tests. Probably name it `test_nbdtr.py`.

File: scipy/special/__init__.py [hunk 1]
```diff
@@ -664,8 +664,10 @@
     spherical_in,
     spherical_kn
 )
+from ._boost import _nbinom_cdf as nbdtrik
 
 __all__ = _ufuncs.__all__ + _basic.__all__ + orthogonal.__all__ + [
+    'nbdtrik',
```
Comment (V0lantis) [inline scipy/special/__init__.py:670]: I put `nbdtrik` here for testing but we might need to change it place to be cleaner.

File: scipy/special/functions.json [hunk 2]
```diff
@@ -1019,11 +1019,6 @@
             "nbdtri": "iid->d"
         }
     },
-    "nbdtrik": {
-        "cdf_wrappers.h": {
-            "cdfnbn2_wrap": "ddd->d"
-        }
-    },
```
Comment (V0lantis) [inline scipy/special/functions.json:1026]: I removed `nbdtrick` from `functions.json` to replace it from the boost implementation. I don't know how all this works, maybe I would be able to directly tell functions.json to call it from `_boost` dir

**Review by V0lantis — COMMENTED** [review]
Thank you @person142 for your help. Nonetheless, it appears that the build fails anyway. I don't really know why since it passes on my computer

**Comment by V0lantis** [comment]
Hello, just a feedback from more digging : 

I resolved some testing errors on my side, but there are unfortunately still a lot. One in particular could be resolved : 

<details>

```bash
_____________________________________________________________________ test_boost[<Data for _nbinom_ppf: negative_binomial_quantile_ipp-negative_binomial_quantile_data>] _____________________________________________________________________
[gw4] darwin -- Python 3.9.6 /Users/arthurvolant/Documents/Development/python/scipy-dev/scipy/venv/bin/python
../../scipy/special/tests/test_data.py:528: in test_boost
    _test_factory(test)
        test       = <Data for _nbinom_ppf: negative_binomial_quantile_ipp-negative_binomial_quantile_data>
../../scipy/special/tests/test_data.py:576: in _test_factory
    test.check(dtype=dtype)
E   AssertionError:
E   Max |adiff|: 0.998533
E   Max |rdiff|: 91.1449
E   Bad results (667 out of 792) for the following points (in output 0):
E              0.09754040092229843               4.285762786865234              0.1269868165254593 =>                            12. !=             11.568381290037562  (rdiff           0.037310207810503114)
E               0.1269868165254593               4.285762786865234              0.1269868165254593 =>                            13. !=             12.977136041273637  (rdiff           0.001761864763815737)
...
____________________________________________________________________ test_boost[<Data for nbdtrik_comp: negative_binomial_quantile_ipp-negative_binomial_quantile_data>] _____________________________________________________________________
[gw4] darwin -- Python 3.9.6 /Users/arthurvolant/Documents/Development/python/scipy-dev/scipy/venv/bin/python
../../scipy/special/tests/test_data.py:528: in test_boost
    _test_factory(test)
        test       = <Data for nbdtrik_comp: negative_binomial_quantile_ipp-negative_binomial_quantile_data>
../../scipy/special/tests/test_data.py:576: in _test_factory
    test.check(dtype=dtype)
E   AssertionError:
E   Max |adiff|: 0.998068
E   Max |rdiff|: 161.724
E   Bad results (631 out of 792) for the following points (in output 0):
E              0.09754040092229843               4.285762786865234              0.1269868165254593 =>                            50. !=              49.67581419477884  (rdiff           0.006526029023903468)
E               0.1269868165254593               4.285762786865234              0.1269868165254593 =>                            47. !=              ...
```

</details>

Those two failing tests are strange because they come from some data which have been generated from boost. But when I am testing on my own computer, I got : 

```c++
#include <iostream>
#include <boost/math/distributions.hpp>

using boost::math::binomial_distribution;
using boost::math::negative_binomial_distribution;
using boost::math::negative_binomial;
// Round up to achieve correct ppf(cdf) round-trips for discrete distributions
typedef boost::math::policies::policy<
        boost::math::policies::discrete_quantile<
                boost::math::policies::integer_round_up > > Policy;


int main() {
    negative_binomial_distribution<long double, Policy> nbdtrik(4.285762786865234, 0.1269868165254593);
    double tmp;

    tmp = boost::math::quantile(nbdtrik, 0.09754040092229843);
    std::cout << tmp << std::endl;

    tmp = boost::math::quantile(complement(negative_binomial(4.285762786865234, 0.1269868165254593), 0.09754040092229843));
    std::cout <<  std::setprecision(20) << tmp;

    return 0;
}
```

gives me : 

`12`, and `50`, and not `11.568381290037562` and `49.67581419477884 ` as in the data in `scipy/special/tests/data/boost.npz`. So I am wondering if there isn't a float precision error here. I just checked out and saw that the data has been generated by @pv in this [PR](https://github.com/scipy/scipy/commit/6c6ded8f552cc4c7ee103e7bf3b2be36041aa67c) (commit : 6c6ded8f552cc4c7ee103e7bf3b2be36041aa67c). I can maybe loose a lot of time trying to generate it again, but I believe @pv, you might have a script that do it automatically with a newer version of Boost?

**Comment by person142** [comment]
A question about the approach here-so `_generate_pyx.py` can already wrap C++ kernels and turn them into ufuncs-see the various Faddeeva-related functions for example:

https://github.com/scipy/scipy/blob/master/scipy/special/functions.json#L376

Why aren't we pointing the existing machinery at the boost kernels? It seems like we are reinventing what `special` already has?

**Comment by V0lantis** [comment]
> It seems like we are reinventing what `special` already has?



> A question about the approach here-`_generate_pyx.py` can already wrap C++ kernels and turn them into ufuncs-see the various Faddeeva-related functions for example:
> 
> https://github.com/scipy/scipy/blob/master/scipy/special/functions.json#L376
> 
> Why aren't we pointing the existing machinery at the boost kernels? It seems like we are reinventing what `special` already has?

Because here we are redefining `boost` implementation of its stats functions (here the negative binomial distribution for example), and not calling directly the c++ functions. I tried multiple alternative using `functions.json` and else, but it is not working as intended. Basically, we generate *cython* files which are calling boost under the hood. So, IMO, here the approach is more like the one we adopted with `_comb.pyx`.

**Comment by person142** [comment]
`comb` is a bit of a special case-it's not a ufunc, and it's not in `cython_special`, so it shouldn't be taken as representative.

> I tried multiple alternative using functions.json and else, but it is not working as intended

I can see how this could but tricky, but don't see that there is a fundamental obstacle here-am I missing something? I don't think maintaining a double ufunc/`cython_special` generation process is the right path forward unless it is impossible to make the existing process work.

**Comment by mckib2** [comment]
I added a specific comment about `Templated_PyUFunc.hpp`, but I think it's more general than that. A lot of the code generation scripts appear to be the same/slightly modified from the version in `scipy.stats._boost`, so moving the common elements out somewhere else would reduce code duplication and help maintainability.  And it looks like the ufuncs you want might already be generated, so if you need the single CDF function from the Boost negative binomial distribution, I would suggest using the ufunc directly from `scipy.stats._boost` and moving the Boost ufunc generation to a common place and having `scipy.stats` and `scipy.special` pull from the same ufunc modules.

**Pull request closed without merging**